rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,202 @@
+an additional element to the `vrl::Value` type. All resources loaded from the 
+config can be added to the compiler variables when it is first invoked.
+
+### Indexing 
+
+Although the first version is not going to be doing any indexing into the data
+it is worth bearing it in mind as we will most likely need to add this in due
+course.
+
+In order to perform the indexing VRL needs to know which fields to index. The 
+criteria is being passed in as an object. If the type def for that object is
+known at runtime, we can extract the fields from this. 
+
+If the type def isn't known we could 
+
+- raise an error 
+- perform the search unindexed
+- create the index at runtime - generally the shape of this object will not 
+  change over time, so most likely the cost of creating the index would only
+  need to be paid on the first event.
+
+The approach we take here should be decided before the first release so we don't
+have to introduce a breaking change further down the line.
+
+Since the resource file is loaded outside of Vrl whilst determining which keys
+need indexing occurs inside Vrl, we will need a way for Vrl to indicate to the 
+resource which indexes need building.
+
+Actual indexing strategies can be decided later.
+
+## Rationale
+
+There is significant customer demand for this feature.
+
+## Drawbacks
+
+1. There are a number of issues with parsing CSV files in particular around
+   handling delimeters and separators.
+2. Performance. Whilst we can ensure all the IO is performed at boot time,
+   searching through that data could still be quite expensive.
+3. Memory use. Since the data file is loaded into memory, the data will use up 
+   some memory. If the file is large enough that could have an impact should
+   Vector be running in a constrained environment.
+
+## Alternatives
+
+### Use a predicate for the search
+
+Instead of using an object to specify the search criteria we could allow the
+user to specify a predicate to determine the row to use for enrichment.
+
+```
+lookup_csv_row("/path/to/file.csv", |row| row.some_key = .some_field)
+```
+
+Note the ability to specify closures this is not yet available in VRL.
+
+The advantage of this is it allows the user to specify a more complex
+set of criteria beyond simple equality. The downside is that it would be very
+hard use indexes to ensure the lookup remains performant.
+
+There is nothing that would prevent us from providing both options.
+
+### Provide other data sources
+
+This RFC proposes using a CSV file as a datasource. It time we may also need to
+source the data from a JSON file, SQL database or Http source.
+
+Using SQL and Http in such a way that would require VRL to perform a lookup
+each event would significantly impact performance and would also require a 
+fairly significant modification to VRL to allow VRL functions to run 
+asynchronously.
+
+## Outstanding Questions
+
+- What approach should we taken around determining the indexed fields?
```

Comment:
Yes, I agree.  It says at the top:

> The first version of this function will not handle any indexing of this data.

But I also think that since it is incredibly likely that we will have to add this, we should consider it in the design from the start to avoid backing us into a corner.

Configuration on the `resources` block is a possibility, and would be essential if we allowed remap to do searches on dynamic fields where we know what they are searching on at boot time.